lib/scanner/xml-scanner/xml-scanner.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2018 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#ifndef XML_SCANNER_H_INCLUDED
+#define XML_SCANNER_H_INCLUDED
+
+#include "syslog-ng.h"
+#include "messages.h"
+
+#include <string.h>
+
+typedef struct _XMLScanner XMLScanner;
+
+typedef void (*PushCurrentKeyValueCB)(const gchar *name, const gchar *value, gssize value_length, gpointer user_data);
+
+typedef gboolean (*StartElement) (XMLScanner *self, const gchar *element_name, const gchar **attribute_names,
+                                  const gchar **attribute_values, GError **error);
+typedef gboolean (*EndElement) (XMLScanner *self, const gchar *element_name, GError **error);
+typedef gboolean (*Text) (XMLScanner *self, const gchar *element_name,
+                          const gchar *text, gsize text_len, GError **error);
+
+typedef struct
+{
+  gboolean strip_whitespaces;
+  GList *exclude_tags;
+} XMLScannerOptions;
+
+typedef struct
+{
+  PushCurrentKeyValueCB push_function;
+  gpointer user_data;
+} PushCurrentKeyValue;
```

Comment:
This can be squashed into the introductor patch.

Reply:
I would have a lot of merge conflicts, which would take time to resolve.
I've received the reorganizing step from a previous review, and even back then it hurt much and I've got a green light to leave the commit.